packages/iov-socket/src/queueingstreamingsocket.ts

```diff
@@ -18,27 +21,36 @@ export class QueueingStreamingSocket {
   private isProcessingQueue = false;
   private timeoutIndex = 0;
   private processQueueTimeout: NodeJS.Timeout | null = null;
+  private connectionStatusProducer: DefaultValueProducer<number | null>;
 
   public constructor(url: string, timeout: number = 10_000) {
     this.url = url;
     this.timeout = timeout;
     this.socket = new StreamingSocket(this.url, this.timeout);
     this.events = this.socket.events;
-    this.connected = this.socket.connected;
+
+    this.connectionStatusProducer = new DefaultValueProducer<number | null>(null);
+    this.connectionStatus = new ValueAndUpdates(this.connectionStatusProducer);
+    this.connected = this.socket.connected.then(() => this.connectionStatusProducer.update(WebSocket.OPEN));
   }
 
   public connect(): void {
+    this.connectionStatusProducer.update(WebSocket.CONNECTING);
     this.socket.connect();
   }
 
   public disconnect(): void {
+    this.connectionStatusProducer.update(WebSocket.CLOSING);
     this.socket.disconnect();
+    this.connectionStatusProducer.update(WebSocket.CLOSED);
```

Comment:
At this point, the websocket is in the CLOSING state since WebSocket.close() does not block.

I suggest using our own, higher level connection status codes, which behave slightly different than a reaw WebSocket. Something like:

- `ConnectionStatus.Unconnected` (default at construct time)
- `ConnectionStatus.Connecting`
- `ConnectionStatus.Connected`
- `ConnectionStatus.Disconnected` (as soon as disconnect() was called, no matter what the implementation does)